Python/ceval.c

```diff
@@ -2476,14 +2476,49 @@ _PyEval_GetBuiltinId(_Py_Identifier *name)
 PyObject *
 PyEval_GetLocals(void)
 {
+    // We need to return a borrowed reference here, so some tricks are needed
     PyThreadState *tstate = _PyThreadState_GET();
      _PyInterpreterFrame *current_frame = _PyThreadState_GetFrame(tstate);
     if (current_frame == NULL) {
         _PyErr_SetString(tstate, PyExc_SystemError, "frame does not exist");
         return NULL;
     }
 
-    PyObject *locals = _PyEval_GetFrameLocals();
+    // Be aware that this returns a new reference
+    PyObject *locals = _PyFrame_GetLocals(current_frame);
+
+    // There are three possibilities here:
+    //   1. locals is NULL - that's an error
+    //   2. locals is a PyFrameLocalsProxy - that means current_frame is
+    //      optimized and we need to return a borrowed reference of a dict.
+    //      In this case, put the reference in current_frame.f_locals which
+    //      will be released when the frame is released.
+    //   3. locals is other mappings - that means current_frame is not
+    //      optimized and we can just decref locals and return it because
+    //      it must have other references
+
+    if (locals == NULL) {
+        return NULL;
+    } else if (PyFrameLocalsProxy_Check(locals)) {
+        PyObject* ret = PyDict_New();
+        if (ret == NULL) {
+            Py_DECREF(locals);
+            return NULL;
+        }
+        if (PyDict_Update(ret, locals) < 0) {
+            Py_DECREF(ret);
+            Py_DECREF(locals);
+            return NULL;
+        }
+        Py_DECREF(locals);
+        Py_XSETREF(current_frame->f_locals, ret);
+        return ret;
+    }
+
+    assert(PyMapping_Check(locals));
```

Comment:
Is this always a `dict` (not just a mapping)? The docs say that this function returns a dictionary:

https://docs.python.org/3/c-api/reflection.html#c.PyEval_GetLocals

Reply:
Unfortunately no. This is probably a neglection in the docs. `exec()` takes [any mapping object](https://docs.python.org/3/library/functions.html#exec) as `locals` and inside that scope, `locals()` will return whatever mapping object that is passed in. We also can't convert it to dict because that would be another disaster. Maybe we should fix the docs? However, in normal cases, `locals()` (and `PyEval_GetLocals`) returns a dict.

However, I need to change more than this. I had to add an extra field in `FrameObject` because of the matter I mentioned above, so this is not the final version :(

Reply:
Yeah, I just partially fixed the docs around this in https://github.com/python/cpython/pull/119893 so they're at least correct for optimized scopes.

As @gaogaotiantian noted, there are still some mentions of `dictionary` in relation to `locals` that aren't strictly true, though (since class scopes aren't always dictionaries)

Reply:
Also see https://github.com/python/cpython/issues/118934#issuecomment-2143330127 in relation to the docs and the ambiguity over the expected behaviour of `PyEval_GetLocals` in optimized scopes.

Reply:
https://github.com/python/cpython/pull/119932 fixed up the lingering mentions of locals always being a dict (as well as aligning the `PyEval_GetLocals` docs with the way things will work once this PR is complete & merged).